app/Exceptions/Handler.php

```diff
@@ -24,4 +35,101 @@ class Handler extends ExceptionHandler
      * @var array
      */
     protected $dontFlash = [];
+
+    public function setResponseFactory(ResponseFactory $factory): self
+    {
+        $this->responseFactory = $factory;
+
+        return $this;
+    }
+
+    public function setDebug(bool $debug): self
+    {
+        $this->debug = $debug;
+
+        return $this;
+    }
+
+    /**
+     * Render an exception into an HTTP response.
+     *
+     * @param Request   $request
+     * @param Throwable $exception
+     * @return JsonResponse
+     */
+    public function render($request, Throwable $exception)
+    {
+        if ($exception instanceof ValidationException) {
+            return $this->responseFactory->json(
+                [
+                    'errors' => $this->mapValidationException($exception),
+                ],
+                $exception->getCode(),
+                [
+                    'Content-Type' => 'application/vnd.api+json',
+                ]
+            );
+        }
+
+        $error = [
+            'status' => $exception->getCode(),
+            'message' => $exception->getMessage(),
+        ];
+
+        if ($this->debug === true) {
+            $error['trace'] = $this->getTrace($exception);
+        }
+
+        return $this->responseFactory->json(
+            [
+                'errors' => [
+                    $error,
+                ],
+            ],
+            $exception->getCode(),
+            [
+                'Content-Type' => 'application/vnd.api+json',
+            ]
+        );
+    }
+
+    private function mapValidationException(ValidationException $exception): array
+    {
+        $validator = $exception->validator;
+        $invalidAttributes = array_keys($validator->failed());
+        $errors = [];
+
+        foreach ($invalidAttributes as $invalidAttribute) {
+            $errorMessages = $validator->errors()->get($invalidAttribute);
+
+            foreach ($errorMessages as $errorMessage) {
+                $pointer = str_replace('.', '/', $invalidAttribute);
+                $title = strpos($errorMessage, 'required') !== false ? 'Missing input' : 'Invalid input';
+                $errors[] = [
+                    'status' => $exception->getCode(),
+                    'message' => $exception->getMessage(),
+                    'title' => $title,
+                    'detail' => $errorMessage,
+                    'source' => ['pointer' => $pointer],
+                ];
+            }
+        }
+
+        return $errors;
+    }
+
+    private function getTrace(Throwable $exception): string
+    {
+        $trace = $exception->getTrace();
+
+        try {
+            if (json_encode($trace) === false) {
+                $trace = 'Trace is not available.';
+            }
+        } catch (Throwable $e) {
+            $trace = 'Trace is not available.';
+        }
+
+        return $trace;
```

Comment:
I believe the following will do the same without the $trace variable
```suggestion
        try {
            if (json_encode($trace) === false) {
                return 'Trace is not available.';
            }
        } catch (Throwable $e) {
            return 'Trace is not available.';
        }

        return $exception->getTrace();
```